pypuf/learner/evolution_strategies/becker.py

```diff
@@ -0,0 +1,311 @@
+                self.num_of_LTFs += 1
+        self.different_LTFs = self.set_pole_of_LTFs(self.instance, self.different_LTFs)
+        return LTFArray(self.different_LTFs, self.transform, self.combiner, bias=False)
+
+    def is_different_LTF(self):
+        # returns True iff new_LTF is different from previously learned LTFs
+        if self.num_of_LTFs == 0 :
+            return True
+        weight_arrays = self.different_LTFs[:self.num_of_LTFs-1, :]
+        new_LTFArray = LTFArray(self.new_LTF, self.transform, self.combiner)
+        different_LTFs = self.build_LTFArrays(weight_arrays)
+        challenges = tools.sample_inputs(self.instance.n, self.challenge_num)
+        responses = np.empty((self.num_of_LTFs, self.challenge_num))
+        for i, challenge in enumerate(challenges):
+            responses[0, i] = new_LTFArray.eval(challenge)
+            for j, current_LTF in enumerate(different_LTFs):
+                responses[j+1, i] = current_LTF.eval(challenge)
+        return not self.is_correlated(responses)
+
+    def learn_LTF(self):
+        # this is the main CMA-ES algorithm like that from Hansen
+        terminate = False
+        while not terminate:
+            challenges = tools.sample_inputs(
+                    self.instance.n, self.challenge_num)
+            measured_rels = np.zeros(self.challenge_num)
+            while np.sum(measured_rels) < (1-self.unreliability)*self.challenge_num:
+                challenges = tools.sample_inputs(
+                    self.instance.n, self.challenge_num)
+                measured_rels = self.measure_rels(
+                    self.instance, challenges, self.challenge_num, self.repeat,
+                    self.epsilon)
+            sorted_individuals = self.fitness(
+                challenges, self.challenge_num, measured_rels,
+                self.individuals, self.epsilon)
+            if self.terminate:
+                break
+            parent = self.get_parent(
+                sorted_individuals, self.parent_size, self.priorities)
+            cm_mu = self.get_cm_mu(
+                sorted_individuals, self.parent_size, self.priorities)
+            self.individuals = self.reproduce(
+                self.mean, self.cov_matrix, self.pop_size, self.step_size)
+            self.mean = self.update_mean(
+                self.mean, self.step_size, parent)
+            self.path_cm = self.cumulation_for_cm(
+                self.path_cm, self.c_c, self.path_ss, self.instance.n,
+                self.mu_w, parent)
+            self.path_ss = self.cumulation_for_ss(
+                self.path_ss, self.c_sigma, self.mu_w, self.cov_matrix, parent)
+            self.cov_matrix = self.update_cm(
+                self.cov_matrix, self.c_1, self.c_mu, self.path_cm, cm_mu)
+            self.step_size = self.update_ss(
+                self.step_size, self.c_sigma, self.d_sigma, self.path_ss)
+        return self.new_LTF
```

Comment:
Try to use less class variables. The heavy usage of class variables in this class makes it hard to follow the data flow. Take this return statement as an example: where does `self.new_LTF` come from? If defined as a local variable, it would be easy to see where it gets assigned. But as a class variable, it could be in any function. Restricting the scope of variables is also a good practice to avoid unintended assignments and hence a good practice to avoid bugs.